static/game/controlsmenu.js

```diff
@@ -0,0 +1,114 @@
+import {
+  buttons,
+  getNextGamepadButton,
+  getNextKey,
+  getUsingKeyboard,
+  saveControls
+} from "../modules/buttons.js";
+import { getScreenDimensions } from "../modules/displaymanager.js";
+import { toggleGuiElement, addToGui } from "../modules/gamemanager.js";
+import { Vector } from "../modules/vector.js";
+import { centeredText, rect } from "./draw.js";
+import { Menu } from "./menu.js";
+import { DirectionalControlMenu } from "./directionalcontrolmenu.js";
+
+export class ControlsMenu extends Menu {
+  constructor() {
+    const screenDimensions = getScreenDimensions();
+    super(new Vector(0, 0), screenDimensions.width, screenDimensions.height);
+    this.itemWidth = 1200;
+    this.textAlign = /** @type {CanvasTextAlign} */ ("left");
+    // make a sub-menu for each directional
+    this.childMenus = [];
+    for (const dir of buttons.getDirectionals()) {
+      const dcm = new DirectionalControlMenu(dir);
+      dcm.active = false;
+      addToGui(dir.name + "controlmenu", dcm);
+      this.childMenus.push(dcm);
+    }
+  }
+
+  action() {
+    const items = [];
+    for (const directional of buttons.getDirectionals()) {
+      items.push({
+        text: directional.name,
+        func: () => {
+          this.active = false;
+          toggleGuiElement(directional.name + "controlmenu");
+        }
+      });
+    }
+
+    for (const button of buttons.getButtons()) {
+      let displayKey = getUsingKeyboard() ? button.key : button.gpButtonIndex;
+      if (displayKey === " ") displayKey = "Space";
+      items.push({
+        text: button.name + "\t" + displayKey,
+        func: () => {
+          if (getUsingKeyboard()) {
+            button.key = "Press a key...";
+            getNextKey().then(newKey => {
+              button.key = newKey;
+              saveControls();
+            });
+          } else {
+            const oldButton = button.gpButtonIndex;
+            // @ts-ignore I know we're assigning a string to a number, but it's
```

Comment:
There has to be a better way. This would be the only time we're using `@ts-ignore`

Reply:
The better way would be a lot more complicated. If we're going to use loosey goosey JavaScript we might as well take advantage of it when we can. I think a simple case like this along with an explanatory comment is clear enough and it doesn't cause any issues.

Reply:
Ya, sure, as long as I'm allowed to fix it.